app.js

```diff
@@ -103,3 +110,46 @@ app.post('/melding', async function (req, res, next) {
     }).end();
   }
 });
+
+const lock = new Lock();
+
+app.post('/download-status-update', async function (req, res) {
+  //The locking is needed because the delta-notifier sends the same request twice to this API because a status update is both deleted and inserted. We don't want this; we can't change that for now, so we block such that no 2 requests are handled at the same time and then limit the way status changes can be performed.
+  await lock.acquire();
+  try {
+    //Because the delta-notifier is lazy/incompetent we need a lot more filtering before we actually know that a resource's status has been set to ongoing
+    const actualStatusChange = req.body
+      .map((changeset) => changeset.inserts)
+      .filter((inserts) => inserts.length > 0)
+      .flat()
+      .filter((insert) => /http:\/\/data.lblod.info\/id\/remote-data-objects\//.test(insert.subject.value))
+      .filter((insert) => insert.predicate.value === env.ADMS_STATUS_PREDICATE)
+      .filter((insert) => insert.object.value === env.DOWNLOAD_STATUSES.ongoing ||
+                          insert.object.value === env.DOWNLOAD_STATUSES.success ||
+                          insert.object.value === env.DOWNLOAD_STATUSES.failure);
+    for (const remoteDataObjectTriple of actualStatusChange) {
+      const { downloadTaskUri, jobUri, oldStatus, submissionGraph, fileUri, errorMsg } = await getTaskInfoFromRemoteDataObject(remoteDataObjectTriple.subject.value);
+      //Update the status also passing the old status to not make any illegal updates
+      await downloadTaskUpdate(submissionGraph, downloadTaskUri, jobUri, oldStatus, remoteDataObjectTriple.object.value, fileUri, errorMsg);
+    }
+    res.status(200).send().end();
+  }
+  catch (e) {
+    console.error(e.message);
+    if (!e.alreadyStoredError)
```

Comment:
I didn't want to change original code if not needed. There was a place where errors were store, but I also wanted to store errors when capturing an error on a higher level. To avoid storing them twice, I added this little hack. This is not a good solution because we shouldn't add properties to objects like the JavaScript built-in Error instance. Can we fix this in a later stage where we do code clean-up and separate 'boiler plate'-stuff for the submission flow in a library?

Reply:
yes